### PR TITLE
[RSDK-7596] setup encoder model

### PR DESCRIPTION
**Pull request by JohnN193**
first pr to add encoders to the revpi module. This PR adds the encoder model, initializes the encoder, and has placeholders for the encoder APIs. 

ticket: https://viam.atlassian.net/browse/RSDK-7596

### CODE REVIEW

File: revolutionpi/digital_interrupts.go [hunk 1]
```diff
@@ -25,10 +25,15 @@ type digitalInterrupt struct {
...
 	inputOffset      uint16
 	enabled          bool
+	isEncoder        bool
 	interruptAddress uint16
 }
 
-func initializeDigitalInterrupt(pin SPIVariable, g *gpioChip) (*digitalInterrupt, error) {
+type diWrapper struct {
+	pin *digitalInterrupt
+}
```
Comment (JohnN193) [inline revolutionpi/digital_interrupts.go:34]: encoders and digital interrupts are configured the same way in the revolution pi, so I tried to setup the code in a way to share the same initialization code. not sure if this is the best way to do it but it appears to be working. 

I had initially tried to make `digitalInterrupt` implement the encoder interface at first, but unfortunately, our interrupts interface and the resource interface both have a `Name` function, and go doesn't let you overload.

File: revolutionpi/digital_interrupts.go [hunk 2]
```diff
@@ -75,16 +80,22 @@ func initializeDigitalInterrupt(pin SPIVariable, g *gpioChip) (*digitalInterrupt
 
 	// check if the pin is configured as a counter
 	// b[0] == 0 means the interrupt is disabled, b[0] == 3 means the pin is configured for encoder mode
-	if b[0] == 0 || b[0] == 3 {
-		return &digitalInterrupt{}, errors.New("pin is not configured as a counter")
+	if b[0] == 0 || (b[0] == 3 && !isEncoder) {
```
Comment (penguinland) [inline revolutionpi/digital_interrupts.go:83]: Do we want to check the other way, too? If `b[0]` indicates a normal interrupt but `isEncoder` is true, is that a problem?

File: revolutionpi/digital_interrupts.go [hunk 3]
```diff
@@ -75,16 +80,22 @@ func initializeDigitalInterrupt(pin SPIVariable, g *gpioChip) (*digitalInterrupt
...
 	return &di, nil
 }
 
+func (di *diWrapper) Value(ctx context.Context, extra map[string]interface{}) (int64, error) {
+	return di.pin.Value()
+}
+
 // Note: The revolution pi only supports uint32 counters, while the Value API expects int64.
-func (di *digitalInterrupt) Value(ctx context.Context, extra map[string]interface{}) (int64, error) {
+func (di *digitalInterrupt) Value() (int64, error) {
```
Comment (penguinland) [inline revolutionpi/digital_interrupts.go:98]: This no longer matches the GRPC interface to get an interrupt's value. I realize there's a layer of indirection between the protobufs and this function where we can throw away the extra data, but it feels like a brittle design (if you want to start doing fancier things in the future, you need to put this line back the way it used to be). What's the reasoning behind changing this?

File: revolutionpi/digital_interrupts.go (hunk 1 above)
Comment (penguinland) [reply revolutionpi/digital_interrupts.go:34]: This github comment is excellent! Make it a comment in the code, too, or else someone is going to mistakenly try to simplify this in 2025.

File: revolutionpi/encoder.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	pin *digitalInterrupt
+}
+
+// EncoderModel is the model triplet for the rev-pi board encoder.
+var EncoderModel = resource.NewModel("viam-labs", "kunbus", "revolutionpi-encoder")
+
+// EncoderConfig is the config for the rev-pi board encoder.
+type EncoderConfig struct {
+	resource.TriviallyValidateConfig
+	Name string `json:"address_name,omitempty"`
```
Comment (penguinland) [inline revolutionpi/encoder.go:30]: :grimacing: What happens if you omit the name? Won't that cause problems using it?

...also, what is an "address_name"? I don't know the names of any addresses on a board; maybe put in a comment explaining where to look these up.

File: revolutionpi/encoder.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	_ resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (encoder.Encoder, error) {
+	svcConfig, err := resource.NativeConfig[*EncoderConfig](conf)
+	if err != nil {
+		return nil, err
+	}
+	devPath := filepath.Join("/dev", "piControl0")
+	devPath = filepath.Clean(devPath)
```
Comment (penguinland) [inline revolutionpi/encoder.go:51]: Mild preference to roll lines 50-51 into a single statement. Keeping them separate makes it look like `devPath` is expected to be mutable and might change again later.

File: revolutionpi/encoder.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	fd, err := os.OpenFile(devPath, os.O_RDWR, fs.FileMode(os.O_RDWR))
+	if err != nil {
+		err = fmt.Errorf("open chip %v failed: %w", devPath, err)
+		return nil, err
+	}
+	gpioChip := gpioChip{dev: devPath, logger: logger, fileHandle: fd}
```
Comment (penguinland) [inline revolutionpi/encoder.go:57]: Mild preference to use a different variable name. This one shadows the previous definition of `gpioChip`, and if we ever want to create another `gpioChip` later in this function, we'll be unable to (that's unlikely, I know, but it's a good habit to get into).

File: revolutionpi/encoder.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	if !enc.isEncoder {
+		return nil, fmt.Errorf("pin %s is not configured as an encoder", name)
+	}
+	return &revolutionPiEncoder{Named: conf.ResourceName().AsNamed(), pin: enc}, nil
+}
+
+func (enc *revolutionPiEncoder) Position(ctx context.Context, positionType encoder.PositionType,
+	extra map[string]interface{},
+) (float64, encoder.PositionType, error) {
+	return 0, encoder.PositionTypeTicks, nil
```
Comment (penguinland) [inline revolutionpi/encoder.go:83]: ...the 0 here is supposed to be something else, right? I don't see how this can work currently.

edit: perhaps this is missing on purpose, to be implemented in a future PR. If so, put in a quick comment explaining that ("TODO: implement this" or similar).

File: revolutionpi/encoder.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func (enc *revolutionPiEncoder) ResetPosition(ctx context.Context, extra map[string]interface{}) error {
+	return nil
```
Comment (penguinland) [inline revolutionpi/encoder.go:87]: I'd return an error saying that resetting the position is not implemented. Returning nil makes it look like we successfully reset the position, which we did not.

edit: perhaps this is missing on purpose, to be implemented in a future PR. If so, put in a quick comment explaining that ("TODO: implement this" or similar).

File: revolutionpi/encoder.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func (enc *revolutionPiEncoder) Properties(ctx context.Context, extra map[string]interface{}) (encoder.Properties, error) {
+	return encoder.Properties{TicksCountSupported: true, AngleDegreesSupported: false}, nil
+}
+
+func (enc *revolutionPiEncoder) DoCommand(ctx context.Context, req map[string]interface{}) (map[string]interface{}, error) {
+	resp := make(map[string]interface{})
+	return resp, nil
```
Comment (penguinland) [inline revolutionpi/encoder.go:96]: Either put in a comment explaining that this will be implemented in an upcoming PR, or else return an error saying that DoCommand is not implemented. The current response makes it look like we successfully did whatever the RPC told us to do, but we didn't actually do that.

File: revolutionpi/encoder.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+	"path/filepath"
+
+	"go.viam.com/rdk/components/encoder"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/resource"
+)
+
+type revolutionPiEncoder struct {
+	resource.Named
+	resource.TriviallyReconfigurable
```
Comment (penguinland) [inline revolutionpi/encoder.go:20]: This resource claims to be trivially reconfigurable, but the config describes the pin name. If someone changes the pin name in their config, aren't we supposed to do something during reconfiguration? I wonder if this line was supposed to be `resource.AlwaysRebuild` instead of `TriviallyReconfigurable`.

File: revolutionpi/encoder.go (hunk 7 above)
Comment (oliviamiller) [inline revolutionpi/encoder.go:83]: Will there be another PR to report position?

File: revolutionpi/digital_interrupts.go (hunk 2 above)
Comment (JohnN193) [reply revolutionpi/digital_interrupts.go:83]: so I added the `di.isEncoder = b[0] == 3` logic to handle this, so anything trying to use an encoder can validate that the pin is an encoder. Wasn't sure if this was better than an explicit error though and I can be convinced to make it an error instead

File: revolutionpi/digital_interrupts.go (hunk 3 above)
Comment (JohnN193) [reply revolutionpi/digital_interrupts.go:98]: pretty much what you said tbh. We weren't using context and extra so I felt like we could simplify the signature. I can see a world where we might want the context, but since `extra` is pretty much only meant to hack in extra methods I did not believe it was necessary to keep it.

File: revolutionpi/encoder.go (hunk 4 above)
Comment (JohnN193) [reply revolutionpi/encoder.go:30]: since the revolution pi has all of the addresses in memory tied to a name I felt like address_name was an okayish way to say that. At first I was just gonna call it name but then it felt weird for a component to have multiple fields called name. pin_name would also work. 

Also wasn't super focused on the name in this pr(mentioned that in another comment) since it technically had not been scoped yet.

File: revolutionpi/encoder.go (hunk 7 above)
Comment (JohnN193) [reply revolutionpi/encoder.go:83]: yeah I put in the description of the PR that api implementations will be handled in a followup pr to try to limit the scope

Comment (JohnN193) [reply revolutionpi/encoder.go:83]: yeah I put in the description of the PR that api implementations will be handled in a followup pr to try to limit the scope

File: revolutionpi/digital_interrupts.go (hunk 3 above)
Comment (penguinland) [reply revolutionpi/digital_interrupts.go:98]: :shrug: I can see a future where you need to undo this change, and can also see a future where this is fine as-is. I trust your judgment; make it how you want.

File: revolutionpi/digital_interrupts.go [hunk 11]
```diff
@@ -102,13 +113,13 @@ func (di *digitalInterrupt) Value(ctx context.Context, extra map[string]interfac
...
 }
 
-func (di *digitalInterrupt) Name() string {
-	return di.pinName
+func (di *diWrapper) Name() string {
+	return di.pin.pinName
 }
 
-func (di *digitalInterrupt) RemoveCallback(c chan board.Tick) {}
+func (di *diWrapper) RemoveCallback(c chan board.Tick) {}
```
[the remark this answers is not in the document]
Comment (JohnN193) [reply revolutionpi/digital_interrupts.go:120]: was looking to wait until [this ticket](https://viam.atlassian.net/browse/RSDK-7337) to start removing these

File: revolutionpi/encoder.go (hunk 4 above)
Comment (JohnN193) [reply revolutionpi/encoder.go]: tbh just changed the name since other parts of the code/readme use "pin name"

File: revolutionpi/encoder.go (hunk 7 above)
Comment (JohnN193) [reply revolutionpi/encoder.go:91]: added a comment

**Review by oliviamiller — APPROVED** [review]

File: revolutionpi/digital_interrupts.go [hunk 12]
```diff
@@ -16,6 +16,12 @@ const (
 	inputModeOffset = 88
 )
 
+// digitalInterrupt is the struct used for configuring an interrupt or encoder.
+// encoders and digital interrupts are configured the same way in the revolution pi.
+// the encoder & digital interrupt interface cannot be satisfied by the same struct due
+// to both interfaces requiring different Name() methods.
+// Go does not allow for overloads, so instead the revolutionPiEncoder and diWrapper structs
+// are used to implement their respective interfaces.
```
Comment (penguinland) [inline revolutionpi/digital_interrupts.go:24]: :+1: This is a really great comment! Thanks

File: revolutionpi/digital_interrupts.go (hunk 2 above)
Comment (penguinland) [reply revolutionpi/digital_interrupts.go:90]: If a user sets up an encoder in their config, but `b[0]` indicates it's not an encoder, we're not going to treat it like one, and the user will be surprised. I'd make it an explicit error. (a doubt: I prefer explicit errors for all kinds of things that other people don't, and I might be overzealous here. You know this code better than I do: if you think this is a bad idea, leave as-is.)

**Review by penguinland — APPROVED** [review]
LGTM!

File: revolutionpi/digital_interrupts.go (hunk 2 above)
Comment (JohnN193) [reply revolutionpi/digital_interrupts.go:90]: nah you right, I was using a later check on isEncoder to see if it was an encoder, but the initialize function should fail if the pin isn't able to initialize correctly
